extra_data/writer.py

```diff
@@ -34,19 +64,24 @@ def _make_control_index(self, source, data_tids):
             firsts = np.zeros_like(counts)
             firsts[1:] = np.cumsum(counts)[:-1]  # firsts[0] is always 0
             self.indexes[source] = (firsts, counts)
-            self.data_sources.add('CONTROL/' + source)
 
-    def add_instrument_dataset(self, source, key):
+    def copy_instrument_dataset(self, source, key):
+        """Copy data into an INSTRUMENT dataset"""
         a = self.data.get_array(source, key)
         path = 'INSTRUMENT/{}/{}'.format(source, key.replace('.', '/'))
-        self.file[path] = a.values
+        self.file[path][:] = a.values
         self._make_instrument_index(source, key, a.coords['trainId'].values)
 
+    def copy_instrument_source(self, source):
+        """Copy data for all keys of one INSTRUMENT source"""
+        for key in self.data.keys_for_source(source):
+            self.copy_instrument_dataset(source, key)
+
     def _make_instrument_index(self, source, key, data_tids):
+        """Create & store an index for an instrument data group"""
         index_path = source + '/' + key.partition('.')[0]
         if index_path not in self.indexes:
             self.indexes[index_path] = self._generate_index(data_tids)
-            self.data_sources.add('INSTRUMENT/' + index_path)
 
     def _generate_index(self, data_tids):
         """Convert an array of train IDs to first/count for each train"""
```

Comment:
You copy yourself here.

These three methods with ***instrument*** in the middle of name and `_generate_index` almost verbatimly repeat previous three methods with ***control*** in the middle. Difference in datasource path, `index_path` and following `assert`. There is one more `assert` in `copy_control_dataset` which looks weired.

Could it better to generalize this code with one or two `if` statements?

Or even implement an academical OOP solution: make `BaseDataset` and two children `InstrumentDataset`, `ControlDataset` and move all this methods to the `BaseDataset`with minimal re-implementation in children. And build `FileWriter.data_sources` set using these classes. Then the FileWriter becames beutifully extandable on image dataset also.

Reply:
Thanks, I've followed your suggestion to combine some methods and separate pieces with `if` statements where necessary. I haven't tried to reorganise it to a more OOP approach at the moment.